### ISSUE TITLE
mocha-allure-reporter does not work properly with Mocha --parallel option

### ISSUE THREAD

**Issue by dprantl03** [allure-framework/allure-mocha#65]
As of v8.0, Mocha supports running spec files in parallel.  When running a simple test file with the --parallel flag passed to the mocha executable (doesn't need to be multiple files), the subsequent generated allure report has all the correct test suites and cases, but:

1) All failed cases are displayed yellow (broken) regardless of whether they are failed or broken
2) All test cases display a run time of 0s

Other Mocha reporting modules have had to make modifications to support the --parallel flag.  Is this going to be implemented for mocha-allure-reporter?

**Comment by dprantl03** [comment]
https://www.npmjs.com/package/mochawesome#parallel-mode

**Comment by just-boris** [comment]
I have investigated this story a bit.

Mocha runs parallel tests in separate processes. Test results are getting buffered and sent back to the parent process upon the suite finish.

Allure provides additional features, which relies on a certain events order, which is different in parallel mode. I checked how mochawesome handles that, they attach additional context to the test object. Mocha serializes that and the reporter can see the data in the parent process.

However, Allure also has attachements feature, which may be very big and will not go through this serizaling logic. They need to be written to the disk right in the test runtime and then anyhow attached to the test result in the parent process.

There are two possible solutions:

1. Attach additional info to the test object and rely on serialization (mochawesome way). This will work, however there will be inaccurate test timestamps and [other limitations](https://github.com/adamgruber/mochawesome#parallel-mode).
2. Patch the test API (`describe` and `it` methods). This way we will have full control on the test flow, however we will need to re-implement some of the mocha logic 

I posted this here as a call for feedback, I want to know what people think, before we get to the implementation of any of the paths.

cc @baev @sseliverstov

**Comment by DJ-Glock** [comment]
Hi all
Is there any chance that it will be fixed? Allure reporter is a great thing but becomes less useful because of this issue :(

**Comment by just-boris** [comment]
Currently in my own projects I am getting around this issue by spawning parallel mocha instances for each test

**Comment by DJ-Glock** [comment]
@just-boris can you please provide me an example?
Do you run N mocha processes by running yarn mocha ... & yarn mocha ... & yarn mocha ...?

**Comment by just-boris** [comment]
**index.ts**

```ts
import glob from 'glob';
import pLimit from 'p-limit';
import prepend from 'prepend-transform';
import execa from 'execa';
import allureGenerate from './allure-generate';

const testFiles = glob.sync(`test/**/*.test.ts`);
const limit = pLimit(MAX_THREADS);

async function run() {
  const failedTests: string[] = [];
  await Promise.all(
    testFiles.map(testFile =>
      limit(async () => {
        const subProcess = execa('node', [require.resolve('./runner'), testFile]);
        const prefix = `${testFile}: `;
        subProcess.stdout?.pipe(prepend(prefix)).pipe(process.stdout);
        subProcess.stderr?.pipe(prepend(prefix)).pipe(process.stderr);
        try {
          await subProcess;
        } catch (error) {
          failedTests.push(testFile);
        }
      })
    )
  );
  await allureGenerate();
  if (failedTests.length > 0) {
    throw new Error('Some tests have been failed: \n' + failedTests.join('\n'));
  }
}

run().catch(error => {
  console.error(error);
  process.exit(1);
});
```

**runner.ts**

```ts
import path from 'path';
import Mocha from 'mocha';

const file = process.argv[2];

export default async function runMocha() {
  const mocha = new Mocha({
    reporter: 'mocha-multi',
    reporterOptions: {
      spec: '-',
      'allure-mocha': {
        stdout: '-',
        options: {
          resultsDir:  'allure-results'
        }
      }
    }
  });

  mocha.addFile(file);

  await new Promise((resolve, reject) =>
    mocha.run(failures => (failures ? reject(new Error('Tests failed')) : resolve()))
  );
}

runMocha().catch(error => {
  console.error(error);
  process.exit(1);
});
```

**Comment by DJ-Glock** [comment]
@just-boris thanks for sharing this.
Unfortunately I'll need to change so many things in my project for this workaround, so not sute If I use it :(

**Comment by jamesmortensen** [comment]
@just-boris Thanks for sharing the two scripts.  I had to change the import syntax to require syntax, and modify some of the arguments to match our setup, but once done, it works great.  To set the allure environment variables, I set this up in a Mocha rootHook beforeEach block.  

On GitHub Actions, our API test runtime went from 18 minutes 56 seconds down to 6 minutes and 43 seconds with 7 worker threads.  

FYI, it seems, with a little bit of polishing, the two scripts could pull arguments from the terminal as a CLI, using something like node-commander.
